### PR TITLE
Make track algo priority order dynamic

### PR DESCRIPTION
**Pull request by makortel**
Currently the "track algo priority order", that is used in track collection merging to dictate the order of iterations, is hardcoded. The order is the one of run2 tracking, but phase1 and phase2 tracking have the iterations in different orders. This PR proposes to make this iteration ordering dynamic via ESProduct `TrackAlgoPriorityOrder` that is configured via tracking eras using `iterativeTkConfig.py`. The `TrackAlgoPriorityOrder` starts with the hardcoded order, and reorders only the algos that are given as parameters.

Tested in CMSSW_9_1_X_2017-03-02-2300. There can be tiny changes in non-run2-tracking in cases the "best algo" gets changed by this reordering and the other track has slightly different parameters. Anything else (run2 tracking, HLT) should have no changes. (and all changes should be caused by the second commit alone, the first one being purely technical)

@rovere @VinInn

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 1]
```diff
@@ -19,9 +19,24 @@
...
+    for producer in producers_by_type(process, "SimpleTrackListMerger", "TrackCollectionMerger", "TrackListMerger"):
+        if not hasattr(producer, "trackAlgoPriorityOrder"):
+            producer.trackAlgoPriorityOrder = cms.string("hltTrackAlgoPriorityOrder")
+    return process
+
 # CMSSW version specific customizations
 def customizeHLTforCMSSW(process, menuType="GRun"):
     # add call to action function in proper order: newest last!
     # process = customiseFor12718(process)
+    process = customiseForXXXXX(process)
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: Please replace, twice, XXXXX by 17771.

Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: Whoops, fixed, thanks.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by silviodonato** [comment]
+1

**Comment by dmitrijus** [comment]
+1

**Comment by perrotta** [comment]
Tiny differences are observed in the phase 1 and phase 2 tracking distribution and none in the run2 ones, as expected.

Jenkins pretends there are also difference in HLT, see e.g. 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_1_X_2017-03-03-1100+17771/18573/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018/HLT__Tracking_ValidationWRTtp_hltIter0PFlowTrackSelectionHighPurity_hltAssociatorByHits_h_phipulleta_Mean.png 
(where it is indeed hard to spot any difference by eyes in those histos, though...)
or
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_1_X_2017-03-03-1100+17771/18573/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D4_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D4+RecoFullGlobal_2023D4+HARVESTFullGlobal_2023D4/Tracking__TrackParameters_highPurityTracks_pt_0to1_GeneralProperties_originalAlgorithm_GenTk.png
(where it looks like one track is actually selected by a different iteration)

@makortel : do you have any explanation for those tiny differences in HLT tracking (as you claimed they were not expected)?

**Comment by slava77** [comment]
@makortel 
do we expect further changes in the order of iterations in phase-1?

**Comment by makortel** [comment]
@slava77 
> do we expect further changes in the order of iterations in phase-1?

No(t for this year at least).

**Comment by makortel** [comment]
@perrotta The first plot (`*_Mean`, also `*_Sigma`) is produced by gaussian fits in the harvesting step. In practice if the input plots for one of them changes, it may induce tiny numerical changes in others that had no changes in their input histograms (IIRC caused by internal state of minuit not being reset between fit calls or something).

The second plot is from offline tracking, where tiny changes in the algo assignment are expected.

**Comment by perrotta** [comment]
+1
Because of https://github.com/cms-sw/cmssw/pull/17771#issuecomment-285374042 and the answer in https://github.com/cms-sw/cmssw/pull/17771#issuecomment-285403729 (indeed, the second plot was from offline tracking, not from HLT...)

**Pull request merged**